velocity-engine-core/src/main/java/org/apache/velocity/runtime/resource/loader/CachingDatabaseObjectsFactory.java

```diff
@@ -0,0 +1,203 @@
+package org.apache.velocity.runtime. resource.loader;
+
+import org.apache.commons.pool2.BaseKeyedPooledObjectFactory;
+import org.apache.commons.pool2.KeyedObjectPool;
+import org.apache.commons.pool2.PooledObject;
+import org.apache.commons.pool2.impl.DefaultPooledObject;
+import org.apache.commons.pool2.impl.GenericKeyedObjectPool;
+import org.apache.commons.pool2.impl.GenericKeyedObjectPoolConfig;
+import org.apache.velocity.util.ExtProperties;
+import org.slf4j.Logger;
+
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+
+/**
+ * <p>Database objects factory which will keep a single connection to be able to cache statements preparation, by means
+ * of appropriate pools.</p>
+ * <p>This class requires the following optional dependency (maven syntax):</p>
+ * <pre><code>
+ *         &lt;dependency&gt;
+ *             &lt;groupId&gt;org.apache.commons&lt;/groupId&gt;
+ *             &lt;artifactId&gt;commons-pool2&lt;/artifactId&gt;
+ *             &lt;version&gt;2.12.0&lt;/version&gt;
+ *             &lt;scope&gt;runtime&lt;/scope&gt;
+ *          &lt;/dependency&gt;
+ * </code></pre>
+ * <p>To use this class, you must add the following property to the example configuration described in
+ * @link{org.apache.velocity.runtime.resource.loader.DataSourceResourceLoader}
+ * </p>
+ * <pre><code>
+ * resource.loader.ds.database_objects_factory.class = org.apache.velocity.runtime.resource.loader.DataSourceResourceLoader<br>
+ * </code></pre>
+ * <p>The default size of each pool of prepared statements (there is one pool per statement) is 50. You can tune it
+ * with:</p>
+ * <pre><code>
+ * resource.loader.ds.database_objects_factory. = org.apache.velocity.runtime.resource.loader.DataSourceResourceLoader<br>
+ * </code></pre>
+ * @see org.apache.velocity.runtime.resource.loader.DataSourceResourceLoader
+ */
+
+public class CachingDatabaseObjectsFactory implements DatabaseObjectsFactory {
+
+    private static final String STATEMENTS_POOL_MAX_SIZE = "statements_pool_max_size";
+    private static final int STATEMENTS_POOL_MAX_SIZE_DEFAULT = 50;
+
+    private DataSource dataSource;
+    private Connection connection;
+    private int poolsMaxSize;
+    private KeyedObjectPool<String, PreparedStatement> statementsPool;
+    protected Logger log = null;
+
+    private class PreparedStatementFactory  extends BaseKeyedPooledObjectFactory<String, PreparedStatement>
+    {
+        @Override
+        public PreparedStatement create(String sql) throws Exception {
+            checkConnection();
+            return connection.prepareStatement(sql);
```

Comment:
Is this operation actually threadsafe? I mean one connection -- one thread, but this create() method will be called by multiple threads in parallel.

Reply:
The checkConnection method is synchronized. IMO that's sufficient.

Reply:
I am talking about `return connection.prepareStatement(sql);`

Reply:
If the connection fails between `checkConnection()` and the following line, I guess we're just out of luck... Or maybe you had something else in mind?

Reply:
Oh, I see what you mean. We cannot assert that the Connection class itself is thread-safe, it depends upon the driver, and we have to assume it is not. So, yes, that is problematic.

I guess it is a common problem for library authors to provide prepared statements caching. We may need something like a weak map with the connection as the key or something...

Reply:
And thus we are reaching a point where the inherent complexity is beyond the scope of our library. So I'm gonna step back, leave in place the factory plugin mechanism but drop the caching factory.